submariner-operator/issue-149-node-placement.md

```diff
@@ -0,0 +1,96 @@
+# Node Placement
+
+[Add support for nodeSelector and toleration scheduling for submariner pods](https://github.com/submariner-io/enhancements/issues/149)
+
+## Summary
+
+Submariner deploys many different agents as part of its core deployment. Gateway, Globalnet Agent and
+Metrics Proxy are deployed on designated gateway nodes, while Route Aents are deployed on all nodes. There is no way
+for users to control or limit where rest of the agent pods are deployed. Some customers want to run these agents on
+specific nodes for billing or networking reasons. Users can't add tolerations directly to these agents coz they'll be
+overwritten by the operator. So, users require an easily configurable way to specify such configurations.
+
+## Proposal
+
+This enhancement proposes adding a new field to `Submariner` and `ServiceDiscovery` for users to configure node
+scheduling. This configuration will then be applied to pods of following submariner agents:
+
+* submariner-operator
+* submariner-lighthouse-agent
+* submariner-lighthouse-coredns
+
+Following agents will be excluded because these are run on Gateway nodes. Users already control which nodes get labeled
+as gateways and ideally we don't want to run anything else on these nodes.
+
+* submariner-gateway
+* submariner-globalnet
+* submariner-metrics-proxy
+
+## Design Details
+
+Add a new `NodePlacement` field to `Submariner` and `ServiceDiscovery` specs.
+
+```Go
+// NodePlacement describes node scheduling configuration for the pods.
+type NodePlacement struct {
+    // NodeSelector defines which Nodes the Pods are scheduled on. The default is an empty list.
+    // +optional
+    NodeSelector map[string]string `json:"nodeSelector,omitempty"`
```

Comment:
Don't we want to use the official type?

https://github.com/kubernetes/api/blob/master/core/v1/types.go#L2801

Reply:
> Don't we want to use the official type?
> 
> https://github.com/kubernetes/api/blob/master/core/v1/types.go#L2801

I used https://github.com/open-cluster-management-io/enhancements/tree/main/enhancements/sig-architecture/58-addon-configuration as referece as that is where the initial use case came from. Okay to go with either, but do lean towards going with official type.

Reply:
Yeah seems the more elaborate terms are used as part of `NodeAffinity`